### DISCUSSION TITLE
[struct_pack] Internal Compile Error

### DISCUSSION THREAD

**Discussion by gnb1007** [alibaba/yalantinglibs#592] [General]
While trying to compile my project using struct_pack in msvc, the compiler gave me a "C1001: Internal compile error", showing file "reflection.hpp" and line "973". This would be the following line of code:

```
template <typename T, typename construct_param_t, typename... Args>
constexpr bool is_constructable=is_constructable_impl<T,construct_param_t,void,Args...>::value;
```

What is the purpose of the above line of code?

I tried reproducing it outside my project, but was not successful. The only two places where I call any struct_pack functions are:

```
template<typename T>
requires Serializable<T>
NetworkSerializationFormat SerializeToNetwork(const T& object) {
	return struct_pack::serialize<NetworkSerializationFormat>(object.GetCurrentState());
}

template<typename T>
requires Serializable<T>
T DeserializeFromNetwork(const NetworkSerializationFormat& object) {
	auto deserialized_object = struct_pack::deserialize<typename T::State>(object);
	return T(deserialized_object.value());
}
```

and the error most likely comes from the second function. Could you give me any help?

(This is my first time posting an Issue in github, I'm sorry if it doesn't qualify as one or if I have been vague, I just really need help)

**Comment by poor-circle** [comment]
We need more info, could you show me the struct `NetworkSerializationFormat`?

**Comment by gnb1007** [comment]
It is just an alias

`using NetworkSerializationFormat = std::string;`

The idea of my code is that each class T will provide a struct type T::State, a constructor that take a T::State and a method that returns a T::State. To ensure this, I created the following concepts:

```
template<typename T>
concept Aggregate = std::is_aggregate_v<std::remove_cvref_t<T>>;

template<typename T>
concept Serializable = Aggregate<typename T::State> && requires(const T & object) {
	typename T::State;
	{ object.GetCurrentState() } -> std::same_as<typename T::State>;
	T(object.GetCurrentState());
};
```

The main class that I am serializing is this one:

```
class GameState {
public:
	HeightMapGenerationParameters height_map_generation_parameters;
	static constexpr double omnibot_max_action_allowance = 5;
	static constexpr int omnibot_extract_capability = 8;
	static constexpr int omnibot_max_mineral_load = 32;
	static constexpr int omnibot_loading_cost = 1;
	static constexpr double omnibot_transporting_multiplier = 1.1;
	static constexpr int players_starting_position_radius = 7;
	static constexpr int mineral_extraction_rate = 8;
	**struct State {
		std::vector<GameCoords> players_starting_position;
		GameMap::State game_map_state;
		std::vector<ControlPod::State> control_pods_states;
		std::vector<Drill::State> drills_states;
		std::vector<Mech::State> mechs_states;
	};**
	GameState(int game_map_width, int game_map_height);
	GameState(const GameState& other);
	~GameState();
	friend void swap(GameState& first, GameState& second);
	GameState& operator=(GameState other);
	**GameState(const State& state);
	State GetCurrentState() const;**
	GameMap* game_map;
	std::vector<GameCoords> players_starting_position;
	std::vector<Entity*> entities;
	std::vector<ControlPod*> control_pods;
	std::vector<Drill*> drills;
	std::vector<Mech*> mechs;
};

```
by using calls such as:
`
received_game_state = DeserializeFromNetwork<GameState>(received);`

Feel free to ask for more info. I could also walk you through the relevant parts of my code if you want that.

**Comment by poor-circle** [comment]
ok, so you want to serailize class GameState?
it's not illegal struct_pack type(it has raw pointer). it will cause a compiler error if compiler no crash

**Reply by gnb1007** [reply]
I want to deserialize a struct GameState::State

**Comment by poor-circle** [comment]
1. you can use unique_ptr instead of raw pointer
2. or you can add your custom function to support it. see https://alibaba.github.io/yalantinglibs/en/struct_pack/struct_pack_intro.html#the-type-cannot-be-abstracted-to-a-type-already-in-the-type-system-1

**Comment by gnb1007** [comment]
I want to deserialize a GameState::State struct, it does not have pointers.

I went back and formatted my posts so the code is clearer, since the template arguments were missing.

**Reply by poor-circle** [reply]
ok, try to add this code after class GameState{…}; to workaround:
```cpp
struct_pack_refl(GameState::State,players_starting_position, game_map_state……);//add all the member name，here
```

**Comment by poor-circle** [comment]
or,show me all the define of the type in GameState::State, such as Gamecoords, GameMap::State. if you can give me a simple reproduce code, i will try to find why compiler crashed.

**Comment by gnb1007** [comment]
I found out that my GameCoords class was not an aggregate type (I should be using GameCoords::State instead, which is a struct with just 2 ints as fields). That is probably why it didn't work.

As an aside, is struct_pack able to support nestd vectors? Something like std::vector<std::vector<int>>? Or, for that matter, any nested stl containers...?

**Reply by poor-circle** [reply]
nested vector is ok, you can get more info from https://alibaba.github.io/yalantinglibs/en/struct_pack/struct_pack_type_system.html

**Comment by gnb1007** [comment]
I still have problems, now with nested vectors of bool. I was able to conjure up a minimum reproducible example:

```
#include "ylt/struct_pack.hpp"

class A {
public:
	struct State {
		std::vector<std::vector<bool>> As;
	};
	A(const State& a) : As(a.As) {

	}

	State GetCurrentState() const {
		return State{ As };
	}

private:
	std::vector<std::vector<bool>> As;
};

template<typename T>
T DeserializeFromNetwork(const std::string& object) {
	auto deserialized_object = struct_pack::deserialize<typename T::State>(object);
	return T(deserialized_object.value());
}

int main() {
	std::string s = "";
	DeserializeFromNetwork<A>(s);
}
```

This gives me the following error:

Error C2664'struct_pack::err_code struct_pack::detail::unpacker<struct_pack::detail::memory_reader,0,false>::deserialize_one<2,18446744073709551615,true,0,std::_Vb_reference<std::_Wrap_alloc<std::allocator<std::_Vbase>>>>(T &)': cannot convert argument 1 from 'std::_Vb_reference<std::_Wrap_alloc<std::allocator<std::_Vbase>>>' to 'T &'

(I didn't know if I should have opened another thread, so I am responding in this one anyway. But the problem is other)

**Reply by poor-circle** [reply]
we dont support `vector<bool>` now. `vector<bool>` is not vector

**Reply by gnb1007** [reply]
Sorry, I don't understand. You said before that nested vectors should be ok, but now you don't support it?

**Reply by poor-circle** [reply]
i mean `vector<bool>`

**Comment by poor-circle** [comment]
The `vector<bool>` is not a STL container, maybe we will support it later. Even we support it, the serialization/deserialization of `vector<bool>` will be slow.

**Reply by poor-circle** [reply]
Add issue https://github.com/alibaba/yalantinglibs/issues/593

**Reply by gnb1007** [reply]
Ah ok, thank you. Was able to make it work, thanks so much for the library and the prompt feedback!
